### ISSUE TITLE
Refactoring: Move string to resx file

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#16624]
This could be extremely useful as changing resx and use that in the code is not even near productive.

**Comment by Cosifne** [comment]
My initial design of this refactoring (WIP, suggestions welcome) Thanks Cheryl's previous notes

How to trigger:
When the user selects a string
1. literal
2. Interpolation
3. Verbatim
(Should some method like string.Format() also be checked?)
(Do I miss some other usage of string in C#?)

Code Action:
1. (Necessarily) Move the string to .resx file 
2. (Potential function may apply here) Let user can create a new resource file to move the string.
(Should we enable user to do this? Since resources file could be added in VS by creating new item, do we need to let them have such short-cut? Also we need to think about how to interact with user to get the file name and the file path)

Problems:
1. How should Interpolation look like in the resource file? if it is
`string h = "Hello";
string w = $"{h} world";` 
and the user wants to move 'w' to the resource, should it be "Hello world" in .resx file or "{0} world"? 

If it is `Hello world`, then should the varibles in the interpolation also needs to be literal(then it should be recusively literal)? Will this behavior creates a lot of false negative situations? (Some users like us often try to put string like "{0} world" into resource files)

If it is "{0} world", then how do we let user know it? (via preview? do we support preview file in .resx file? then we should preview xml file? Or the designer file?)
Also this may create a lot of false positive situations? (Since we will provide this refactoring everytime user selects on a string interpolation, but I think it's OK, since refactoring may not always useful)

**Comment by alrz** [comment]
>If it is "{0} world", then how do we let user know it? 

The resulting expression will be `string.Format(0_world, h)` (perhaps with a renaming annotation on `0_world` for further editing).

**Comment by alrz** [comment]
Also in case there's more than one resx file in the project (or solution?), I think there should be an option for user to select (or create a new one?)

**Comment by Cosifne** [comment]
> 
> 
> Also in case there's more than one resx file in the project (or solution?), I think there should be an option for user to select (or create a new one?)

I am feeling there are many design issues need to be discussed with ide teams, problems you mentioned are something I really want to discuss. I will update this thread when there is anything new.

**Comment by alrz** [comment]
In Roslyn itself, you might want add a resource to another project that we reference, (e.g. we reference common strings from Features project in CSharpFeatures). so considering "solution" (at least those projects that we reference) would really help here. thanks

**Comment by anli-xsigns** [comment]
I think starting with this could be really useful. Maybe in several steps:
1. Move only complete strings to project resx file
2. Move strings like `string h = "Hello"; string w = $"{h} world";` to resx file. I would vote for a style like "{0} world"
3. Make it decidable in dropdown where to move resource (project resx or "local" resx)

Following the pattern "If a problem is too big to be solved, divide it into smaller problems that could be solved." :-)

**Comment by CyrusNajmabadi** [comment]
one thing that makes this challenging is that Roslyn itself has no concept of the resx file.

I would prefer that *first* we make it so that .resx files are representd in teh Roslyn Workspace model.  *Then* we can add features around it.  We would add it in a manner similar to how we added 'additional documents' and how @jasonmalinowski is now adding '.editorconfig' files to the model.

@jasonmalinowski This is an example of why i'd like a much cleaner way to represent different sorts of files in Roslyn.  With the current approach (as we've seen with .editorconfig) it's incredibly costly to add/track a new file concept).  If we rejiggered how we support different doc types, it could then be much easier to add support for this, and then adding a feature like this would likely be much easier.

**Comment by JakeRadMSFT** [comment]
We'd like a "Move All Strings in File/Folder/Project to Resource File.

**Comment by tmat** [comment]
@CyrusNajmabadi Couldn't EmbeddedResource items be added to additional files? Why do we need a different kind of documents?

**Comment by CyrusNajmabadi** [comment]
> Why do we need a different kind of documents?

IN this case, because i think resx *should* have meaning to roslyn.  i.e. it shouldn't just be something inert.  like with .editorconfig these are files with real information in them that i'd like roslyn to expose in a first-class fashoin through the workspace, not just as a link the doc and basically a big blob of text :)

**Comment by tmat** [comment]
It can still have a meaning to the IDE even if it is in additional files. 
You can imagine all sorts of DSLs that are used during build to generate some source file that gets included in the compilation. Like e.g. XAML files. Why would resx be special? I can easily imagine a code fix that updates XAML file.

**Comment by CyrusNajmabadi** [comment]
>  I can easily imagine a code fix that updates XAML file.

Yes.  Xaml files should not be a .AdditionalFile either :)  

Note: i don't think it is.  When we did the work with the xaml team, we added it as an actual real roslyn content type.  This allows them to actually use all our 'Workspace/Document' infrastructure and have it light-up on xaml files.  For example, they can use this to provide lightbulbs without having to reinvent everything that roslyn built :)

**Comment by CyrusNajmabadi** [comment]
> Why would resx be special?

I don't want it to be special.  That's my point.  I think that supporting different content types and having that flow through Roslyn (and to all our services) should be natively supported.  We technically entirely support this concept.  It's just not extensible.  We have to build the support for each new language one at a time.

**Comment by mgoertz-msft** [comment]
> Yes. Xaml files should not be a .AdditionalFile either :)

Timely conversation as I'm about to start looking at our XAML integration with Roslyn again. I have actually been getting mixed messages about this, and it was suggested to me that XAML probably should be part of additional files.

FYI, currently XAML files only exist as regular Documents in a "shadow project" with the XAML content type so that we can take advantage of DocumentAnalyzers and other services for renaming and remove/sort of usings (xmlns in XAML). The main reason for that is that ContentTypes are per project and not per document.

Another limitation is that those XAML documents cannot support syntax trees or semantic models, which limits us to DocumentAnalyzers and also prevents creation of any 3rd party analyzers and code fixes.

But if regular Documents should remain reserved for C#/VB then perhaps the concept of additional files needs to be expanded to allow factories to create specific additional files by ContentType to...

1. Expose access to their own extensibility APIs
2. Leverage other Roslyn language services

**Comment by CyrusNajmabadi** [comment]
> But if regular Documents should remain reserved for C#/VB

Regular docs are not reserved for those two languages. TS and Js are also normal docs.

**Comment by CyrusNajmabadi** [comment]
All a Document means is: I support features. 

> Leverage other Roslyn language 

This is why Document exists :-).   For example, if you want to support classification, or outlining, or completion, or etc. etc. in Roslyn, you do it by exposing Documents.

**Comment by mgoertz-msft** [comment]
Can such documents exist in the same project as C# and VB as is the case with XAML and RESX for example? And still support all these services?

And how would we let those documents expose their own syntax trees (and semantic models)?

**Comment by sharwell** [comment]
@CyrusNajmabadi No more document types.

**Comment by CyrusNajmabadi** [comment]
Yes. I would want this to work. Individual documents can specify their languages, so I would want services to work at that level. 

Wrt syntax/semantics. You should be able to expose that however makes sense to you. Likely the best bet would be extension methods. It's slightly unfortunate that C#/vb are blessed here. I would have preferred they just be exposed through services/extensions. However the initial design there shouldn't limit future language work  imo :-)

**Comment by mgoertz-msft** [comment]
> @CyrusNajmabadi No more document types. They are a huge pain to maintain.

> Yes. I would want this to work. Individual documents can specify their languages, so I would want services to work at that level.

So how do we consolidate these two statements? It would seem to me that making services work at the document level would make what @sharwell said even worse because it would most likely encourage more document types.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi No more document types. They are a huge pain to maintain.

Would need more information behind that statement.

**Comment by CyrusNajmabadi** [comment]
> So how do we consolidate these two statements? 

My preference is simply that languages like C#/VB are not blessed or deeply baked in to the core roslyn kernel (though some methods on Project/Document may always necessitate some break in that goal).

Languages have a clearly delineated API for injecting themselves into the roslyn kernel and having features light up.  You go to the language-specific-api itself for language specific concepts (like symbols/semantics).  

Roslyn then serves as the mechanism for managing structure and the editing experience.  With all the built-in core pieces for orchestration and management of things like remoting, storage/persistence, feature-lightup, UI experiences, etc.  And languages themselves are responsible for answering language specific questions in support of that (like: "what errors do you have in this document?" "what are your classificaitons for this span").  

In other words, language smarts come in through some API similar (or exactly equivalent) to LSP.  But everything else is non-language-specific support for core experiences (i.e. "how does rename behave?" "how do codefix previews get shown").  i.e. all the things you want to share across all/most languages.

**Comment by tmat** [comment]
@mgoertz-msft I don't think we have an agreement on a design that we can share with partners at this time. 

I think two concepts are being conflated on this thread:
1) TypeScript/F# using Roslyn infrastructure and creating projects with .ts and .fs files. 
2) resx, XAML, and other files that indirectly contribute to the compilation of the project

I believe these are fundamentally different. 

Files in [2] are arbitrary files (could be any format whatsoever, including `.txt` files) that are processed by a custom build step preceding csc/vbc. The result of the processing is C#/VB file passed to the compiler (and potentially to any analyzer that is interested in the content).

[1] are represented in Documents, [2] are represented in AdditionalDocuments. 

> This is why Document exists :-). For example, if you want to support classification, or outlining, or completion, or etc. etc. in Roslyn, you do it by exposing Documents.

We are moving away from this model though. The goal is to remove the use of Roslyn infrastructure for the features you mention here and provide the features from LSP servers of the respective languages.

Features like Find All References could work across Documents and AdditionalDocuments linking the symbols found in Documents with the entities in AdditionalDocuments that originated them.

**Comment by mgoertz-msft** [comment]
@tmat Thank you. I would argue though that XAML files are a bit different than other additional files in that the elements and attributes in XAML represent real types and properties. In fact, in our XAML language service these XAML elements and attributes are backed by the exact same C# and VB Roslyn symbols.

In other words, XAML is closely tied to the C#/VB Roslyn workspace and therefore putting XAML into a standalone out-of-proc LSP server would not make much sense.

Furthermore, even if completion, outlining, etc. were to come through LSP there are some other benefits of being a Document that we currently rely on, such as previews of Document changes for CodeFixes for example. I don't know what that would look like for AdditionalDocuments.

**Comment by tmat** [comment]
@mgoertz-msft They are actually not that different. For example, entries in resx file (XML nodes) correspond to generated C# properties. In Roslyn we have a custom XML format that we generate C#/VB syntax node types from. XML nodes in these files correspond to real C#/VB types and properties as well.

> , such as previews of Document changes for CodeFixes for example. I don't know what that would look like for AdditionalDocuments.

We would have the same preview of changes for AdditionalDocuments as well.

**Comment by CyrusNajmabadi** [comment]
If all documents are effectively 'additional documents' then what real value is there in the split?  You already need some sort of 'kind' to distinguish those documents so you light up the right features (if that's using lsp or some other mechanism). So at that point, why not just use the existing system we have that already works for this?  Namely documents and language kind?

**Comment by tmat** [comment]
The difference between `Documents` and `AdditionalDocuments` is that Documents match syntax trees in the compilation and are of the language of the project. `AdditionalDocuments` are arbitrary extra files that we we track (flow them from msbuild to Roslyn, etc.).

**Comment by CyrusNajmabadi** [comment]
Closing out as external.  This would be for the resx owners to do anyways given that they know the format of their files and own the code-generaiton from it.  If they want to make their own analyzer/refactoring/fixer to go from code to resx, that would be nice.  But on them.
